src/inmanta/agent/forking_executor.py

```diff
@@ -710,13 +753,40 @@ async def stop_for_agent(self, agent_name: str) -> list[MPExecutor]:
         await asyncio.gather(*(child.stop() for child in children))
         return children
 
-    async def cleanup_inactive_executors(self) -> None:
-        now = datetime.datetime.now().astimezone()
+    def start_periodic_executor_cleanup(self) -> None:
+        async def cleanup_inactive_executors() -> None:
+            """
+            This task cleans up idle executors and reschedules itself
+            """
 
-        for _executor in self.executor_map.values():
-            if now - _executor.connection.last_used_at > timedelta(seconds=self.executor_retention_time):
-                LOGGER.info(
-                    f"Stopping executor {_executor.executor_id.identity()} because it was inactive for longer "
-                    f"than the configured allowed idling time."
-                )
-                await _executor.stop()
+            now = datetime.datetime.now().astimezone()
+            reschedule_interval: float = self.executor_retention_time
+            for _executor in self.executor_map.values():
+                if _executor.can_be_cleaned_up(self.executor_retention_time):
+                    LOGGER.debug(
+                        "Stopping executor %s because it "
+                        "was inactive for %d s, which is longer then the retention time of %d s.",
+                        _executor.executor_id.identity(),
+                        _executor.connection.get_idle_time().total_seconds(),
+                        self.executor_retention_time,
+                    )
+                    async with self._locks.get(_executor.executor_id.identity()):
+                        try:
+                            await _executor.stop()
+                        except Exception:
+                            LOGGER.debug(
+                                "Unexpected error during executor %s cleanup:", _executor.executor_id.identity(), exc_info=True
+                            )
+                else:
+                    reschedule_interval = min(
+                        reschedule_interval,
+                        (
+                            now - _executor.connection.last_used_at + timedelta(seconds=self.executor_retention_time)
+                        ).total_seconds(),
+                    )
+
+            LOGGER.debug(f"Scheduling next cleanup_inactive_executors in {reschedule_interval} s.")
+            await asyncio.sleep(reschedule_interval)
```

Comment:
What if the setting changes? i.e. if it is set first to 1 day, then changed to 1 second, but we're already sleeping for a day, there will be a day's delay.

Reply:
Added a watcher task that periodically checks the setting's value and reschedules the cleanup if it changed

Reply:
As discussed with @wouterdb, this is an agent setting i.e. not currently possible to be updated on the fly while the agent is running.